### PR TITLE
[Enhance] Random generator refactor

### PR DESCRIPTION
**Pull request by shijianjian**
#### Changes
This PR refactored the random generator module, mainly for dragging all param validation, distribution generation into ```__init__``` which reduced the unnecessary object reinitialization.

#### Benchmark

Here we benchmarked with Torchvision default CPU augmentation and Kornia GPU augmentation, on Google Colab K80 GPU with different batch sizes. unit=ms, and ```bs``` means batch size.

The results recorded mean per-sample augmentation speed. Kornia got a better results when larger batch size.

  ops  | torchvision | kornia (bs=1)
 ---- |---- |---- |
RandomPerspective | 4.88±1.82 | 4.74±2.84 |
ColorJitter | 4.40±2.88 | 4.14±3.85 |
RandomAffine | 3.12±5.80 | 3.01±7.80 |
RandomVerticalFlip | 0.32±0.08 | 0.35±0.82 |
RandomHorizontalFlip | 0.32±0.08 | 0.31±0.59 |
RandomRotate | 1.82±4.70 | 1.58±4.44 | 
RandomCrop | 4.09±3.41 | 3.84±3.07 | 
RandomErasing | 2.31±1.47 | 2.32±3.31 |
RandomGrayscale | 0.41±0.18 | 0.45±1.20 |
RandomResizedCrop | 4.23±2.86 | 4.07±2.67 |
RandomCenterCrop | 2.93±1.29 | 2.88±2.34 |

  ops  | kornia (bs=128) | THIS_PR (bs=128) | kornia (bs=32) | THIS_PR (bs=32)
 ---- |---- |---- |---- |---- |
RandomPerspective | 0.19±2.31 | 0.20±27.00 | 0.35±2.06ms | 0.37±2.67ms
ColorJitter | 0.84±12.85 | 0.83±12.96 | 0.89±17.12ms | 0.90±24.68ms
RandomAffine | 0.18±7.34 | 0.18±6.30 | 0.30±4.23ms | 0.30±4.39ms
RandomVerticalFlip | 0.01±0.49 | 0.01±0.35 | 0.02±0.11ms | 0.02±0.13ms
RandomHorizontalFlip | 0.01±2.29 | 0.01±0.37 | 0.02±1.00ms | 0.01±0.26ms
RandomRotate | 0.18±28.17 | 0.17±5.69 | 0.25±3.33ms | 0.25±2.09ms
RandomCrop | 0.07±2.76 | 0.08±9.42 | 0.17±6.32ms | 0.16±1.17ms
RandomErasing | 0.56±7.02 | 0.57±9.74 | 0.45±4.78ms | 0.44±2.82ms
RandomGrayscale | 0.03±0.10 | 0.03±7.10 | 0.03±0.13ms | 0.03±0.11ms
RandomResizedCrop | 0.13±6.88 | 0.13±8.04 | 0.23±4.02ms | 0.23±5.27ms
RandomCenterCrop | 0.07±4.98 | 0.07±9.41 | 0.13±2.55ms | 0.13±2.20ms

### CODE REVIEW

File: test/augmentation/test_random_generator.py [hunk 1]
```diff
@@ -83,62 +83,54 @@ class TestColorJitterGen(RandomGeneratorBaseTests):
...
+                torch.as_tensor(brightness if brightness is not None else torch.tensor([0., 0.]), device=device, dtype=dtype),
+                torch.as_tensor(contrast if contrast is not None else torch.tensor([0., 0.]), device=device, dtype=dtype),
+                torch.as_tensor(saturation if saturation is not None else torch.tensor([0., 0.]), device=device, dtype=dtype),
+                torch.as_tensor(hue if hue is not None else torch.tensor([0., 0.]), device=device, dtype=dtype),
+            )(torch.Size([8]))
 
     def test_random_gen(self, device, dtype):
         # TODO(jian): crashes with pytorch 1.10, cuda and fp64
         if torch_version_geq(1, 10) and "cuda" in str(device):
             pytest.skip("AssertionError: Tensor-likes are not close!")
```
Comment (shijianjian) [inline test/augmentation/test_random_generator.py:125]: To investigate.

**Comment by shijianjian** [comment]
@edgarriba We need to take flake8 down to < 4.0, check this: https://github.com/PyCQA/flake8/issues/1419

File: docs/source/augmentation.rst [hunk 2]
```diff
@@ -14,6 +14,42 @@ in bounding boxes or landmark keypoints for object detection tasks. It gives the
 perform complex data augmentations pipelines.
 
 
+Benchmark
+---------
+
+.. table:: Here is a benchmark performed on Google Colab K80 GPU with different libraries and batch sizes. This benchmark shows
+   strong GPU augmentation speed acceleration brought by Kornia data augmentations. The image size is fixed to 224x224 and the
+   unit is milliseconds (ms).
```
Comment (edgarriba) [inline docs/source/augmentation.rst:22]: we should provide the link to that colab script to reproduce

**Review by edgarriba — COMMENTED** [review]
can we improve in this PR the interface to reuse parameters between calls ? I think it's something that's continuously asked.

**Comment by shijianjian** [comment]
@edgarriba Please approve and merge. Tests failed on flake8 bugs.

**Review by edgarriba — COMMENTED** [review]
What’s the exact issue with flake here ? We can eventually fix version but better to find the reason

**Comment by shijianjian** [comment]
> @edgarriba We need to take flake8 down to < 4.0, check this: [PyCQA/flake8#1419](https://github.com/PyCQA/flake8/issues/1419)

This is the issue.

**Comment by edgarriba** [comment]
@shijianjian then downgrade flake version

File: kornia/augmentation/base.py [hunk 3]
```diff
@@ -65,7 +82,10 @@ def transform_tensor(self, input: torch.Tensor) -> torch.Tensor:
         raise NotImplementedError
 
     def generate_parameters(self, batch_shape: torch.Size) -> Dict[str, torch.Tensor]:
-        return {}
+        if self._param_generator is not None:
+            return self._param_generator(batch_shape, self.same_on_batch)
+        else:
```
Comment (edgarriba) [inline kornia/augmentation/base.py:87]:
```suggestion
```

File: kornia/augmentation/base.py [hunk 4]
```diff
@@ -65,7 +82,10 @@ def transform_tensor(self, input: torch.Tensor) -> torch.Tensor:
...
+            return {}
```
Comment (edgarriba) [inline kornia/augmentation/base.py:88]:
```suggestion
        return {}
```

File: kornia/augmentation/base.py [hunk 5]
```diff
@@ -26,6 +28,11 @@ class _BasicAugmentationBase(nn.Module):
     By default, the random computations will be happened on CPU with ``torch.get_default_dtype()``.
     To change this behaviour, please use ``set_rng_device_and_dtype``.
 
+    For automatically generating the corresponding ``__repr__`` with full customized parameters, you may need to
+    implement ``_param_generator`` by inheriting ``RandomGeneratorBase`` for generating random parameters and
+    put all static parameters inside ``self.flags``. You may take the advantage of ``PlainUniformGenerator`` to
+    generate simple uniform parmeters with less boilertemplate code.
```
Comment (edgarriba) [inline kornia/augmentation/base.py:34]:
```suggestion
    generate simple uniform parameters with less boilerplate code.
```

File: kornia/augmentation/base.py [hunk 6]
```diff
@@ -137,7 +163,7 @@ class _AugmentationBase(_BasicAugmentationBase):
     Advanced augmentation base class with the functionality of transformation matrix calculations.
 
     Args:
-        pprobability for applying an augmentation. This param controls the augmentation probabilities
+        p:probability for applying an augmentation. This param controls the augmentation probabilities
```
Comment (edgarriba) [inline kornia/augmentation/base.py:166]:
```suggestion
        p: probability for applying an augmentation. This param controls the augmentation probabilities
```

File: kornia/augmentation/augmentation.py [hunk 7]
```diff
@@ -467,6 +447,12 @@ class RandomPerspective(GeometricAugmentationBase2D):
         tensor([[[[0.0500, 0.0961, 0.0000],
                   [0.2011, 0.3144, 0.0000],
                   [0.0031, 0.0130, 0.0053]]]])
+
+    To apply the exact augmenation again, you may take the advantage of the previous parameter state:
+        >>> input = torch.randn(1, 3, 32, 32)
+        >>> aug = RandomPerspective(0.5, p=1.)
+        >>> (aug(input) == aug(input, params=aug._params)).all()
```
Comment (edgarriba) [inline kornia/augmentation/augmentation.py:454]: How’s it working right now. Are the parameters generated before nowing the input shape?

File: kornia/augmentation/augmentation.py [hunk 8]
```diff
@@ -821,12 +717,13 @@ def compute_transformation(self, input: torch.Tensor, params: Dict[str, torch.Te
     def apply_transform(
         self, input: torch.Tensor, params: Dict[str, torch.Tensor], transform: Optional[torch.Tensor] = None
     ) -> torch.Tensor:
-        if self.cropping_mode == 'resample':  # uses bilinear interpolation to crop
+        if self.flags['cropping_mode'] == 'resample':  # uses bilinear interpolation to crop
```
Comment (edgarriba) [inline kornia/augmentation/augmentation.py:720]:
```suggestion
        if self.flags["cropping_mode"] == "resample":  # uses bilinear interpolation to crop
```

File: kornia/augmentation/augmentation.py [hunk 9]
```diff
@@ -516,7 +483,8 @@ def apply_transform(
         _, _, height, width = input.shape
         transform = cast(torch.Tensor, transform)
         return warp_perspective(
-            input, transform, (height, width), mode=self.resample.name.lower(), align_corners=self.align_corners
+            input, transform, (height, width), mode=self.flags['resample'].name.lower(),
```
Comment (edgarriba) [inline kornia/augmentation/augmentation.py:486]:
```suggestion
            input, transform, (height, width), mode=self.flags["resample"].name.lower(),
```

File: kornia/augmentation/augmentation.py [hunk 10]
```diff
@@ -516,7 +483,8 @@ def apply_transform(
...
+            align_corners=self.flags['align_corners']
```
Comment (edgarriba) [inline kornia/augmentation/augmentation.py:487]:
```suggestion
            align_corners=self.flags["align_corners"]
```

File: kornia/augmentation/augmentation.py [hunk 11]
```diff
@@ -847,14 +744,14 @@ def inverse_transform(
         size: Optional[Tuple[int, int]] = None,
         **kwargs,
     ) -> torch.Tensor:
-        if self.cropping_mode != 'resample':
+        if self.flags['cropping_mode'] != 'resample':
```
Comment (edgarriba) [inline kornia/augmentation/augmentation.py:747]:
```suggestion
        if self.flags['cropping_mode'] != "resample":
```

File: kornia/augmentation/augmentation.py [hunk 12]
```diff
@@ -847,14 +744,14 @@ def inverse_transform(
...
             raise NotImplementedError(
-                f"`inverse` is only applicable for resample cropping mode. Got {self.cropping_mode}."
+                f"`inverse` is only applicable for resample cropping mode. Got {self.flags['cropping_mode']}."
```
Comment (edgarriba) [inline kornia/augmentation/augmentation.py:749]:
```suggestion
                f"`inverse` is only applicable for resample cropping mode. Got {self.flags["cropping_mode"]}."
```

File: kornia/augmentation/augmentation.py [hunk 13]
```diff
@@ -847,14 +744,14 @@ def inverse_transform(
...
             )
         if size is None:
             size = self.size
-        mode = self.resample.name.lower() if "mode" not in kwargs else kwargs['mode']
-        align_corners = self.align_corners if "align_corners" not in kwargs else kwargs['align_corners']
+        mode = self.flags["resample"].name.lower() if "mode" not in kwargs else kwargs['mode']
+        align_corners = self.flags["align_corners"] if "align_corners" not in kwargs else kwargs['align_corners']
```
Comment (edgarriba) [inline kornia/augmentation/augmentation.py:754]:
```suggestion
        align_corners = self.flags["align_corners"] if "align_corners" not in kwargs else kwargs["align_corners"]
```

File: kornia/augmentation/augmentation.py [hunk 14]
```diff
@@ -922,21 +825,13 @@ def __init__(
...
     ) -> None:
         super().__init__(p=p, return_transform=return_transform, same_on_batch=same_on_batch, keepdim=keepdim)
-        self._device, self._dtype = _extract_device_dtype([degrees])
-        self.degrees = degrees
-        self.resample: Resample = Resample.get(resample)
-        self.align_corners = align_corners
-        self.flags: Dict[str, torch.Tensor] = dict(
-            interpolation=torch.tensor(self.resample.value), align_corners=torch.tensor(align_corners)
+        self._param_generator = cast(
+            rg.PlainUniformGenerator, rg.PlainUniformGenerator((degrees, 'degrees', 0., (-360., 360.)))
```
Comment (edgarriba) [inline kornia/augmentation/augmentation.py:829]:
```suggestion
            rg.PlainUniformGenerator, rg.PlainUniformGenerator((degrees, "degrees", 0., (-360., 360.)))
```

File: kornia/augmentation/augmentation.py [hunk 15]
```diff
@@ -1046,63 +948,53 @@ def __init__(
...
+                    self.flags["padding"][0],
+                    self.flags["padding"][0]
+                ]
+            elif isinstance(self.flags["padding"], tuple) and len(self.flags["padding"]) == 4:
+                padding = [
+                    self.flags["padding"][3],
+                    self.flags["padding"][2],
+                    self.flags["padding"][1],
+                    self.flags["padding"][0]
+                ]
```
Comment (edgarriba) [inline kornia/augmentation/augmentation.py:983]: Else raise some error ?

File: kornia/augmentation/augmentation.py [hunk 16]
```diff
@@ -1046,63 +948,53 @@ def __init__(
...
-                padding = [self.padding[3], self.padding[2], self.padding[1], self.padding[0]]
-
-        if self.pad_if_needed and shape[-2] < self.size[0]:
-            padding = [0, 0, (self.size[0] - shape[-2]), self.size[0] - shape[-2]]
-
-        if self.pad_if_needed and shape[-1] < self.size[1]:
-            padding = [self.size[1] - shape[-1], self.size[1] - shape[-1], 0, 0]
+        if self.flags["padding"] is not None:
+            if isinstance(self.flags["padding"], int):
+                padding = [self.flags["padding"], self.flags["padding"], self.flags["padding"], self.flags["padding"]]
```
Comment (edgarriba) [inline kornia/augmentation/augmentation.py:969]:
```suggestion
                padding = [self.flags["padding"] for _ in range(4)]
```

Comment (edgarriba) [reply kornia/augmentation/augmentation.py:969]: Maybe you can do similar below

File: kornia/augmentation/random_generator/random_generator.py [hunk 17]
```diff
@@ -1,12 +1,1105 @@
...
+                factor = torch.as_tensor(factor, device=device, dtype=dtype)
+            elif center is None or bound is None:
+                raise ValueError(f"`center` and `bound` should be both None or provided. Got {center} and {bound}.")
+            else:
+                factor = _range_bound(
+                    factor, name, center=center, bounds=bound, device=device, dtype=dtype
+                )
+            self.sampler_dict.update({name: Uniform(factor[0], factor[1], validate_args=False)})
+
+    def forward(self, batch_shape: torch.Size, same_on_batch: bool = False):  # type:ignore
```
Comment (edgarriba) [inline kornia/augmentation/random_generator/random_generator.py:115]: Return type

File: kornia/augmentation/random_generator/random_generator.py [hunk 18]
```diff
@@ -1,12 +1,1105 @@
...
+            self.ksize_sampler = Uniform(self.kernel_size // 2, self.kernel_size // 2, validate_args=False)
+        elif isinstance(self.kernel_size, tuple):
+            # kernel_size is fixed across the batch
+            if len(self.kernel_size) != 2:
+                raise AssertionError(f"`kernel_size` must be (2,) if it is a tuple. Got {self.kernel_size}.")
+            self.ksize_sampler = Uniform(self.kernel_size[0] // 2, self.kernel_size[1] // 2, validate_args=False)
+        else:
+            raise TypeError(f"Unsupported type: {type(self.kernel_size)}")
+
+        self.angle_sampler = Uniform(angle[0], angle[1], validate_args=False)
```
Comment (edgarriba) [inline kornia/augmentation/random_generator/random_generator.py:871]: Is it possible to customise the different samplers ? Not only here

File: kornia/augmentation/random_generator/random_generator3d.py [hunk 19]
```diff
@@ -1,13 +1,594 @@
...
+        if size.shape != torch.Size([batch_size, 3]):
+            raise AssertionError(
+                "If `size` is a tensor, it must be shaped as (B, 3). "
+                f"Got {size.shape} while expecting {torch.Size([batch_size, 3])}."
+            )
+        if not (
+            isinstance(depth, (int,)) and isinstance(height, (int,)) and isinstance(width, (int,))
+            and depth > 0 and height > 0 and width > 0
+        ):
+            raise AssertionError(f"`batch_shape` shoud not contain negative values. Got {(batch_shape)}.")
```
Comment (edgarriba) [inline kornia/augmentation/random_generator/random_generator3d.py:285]:
```suggestion
            raise AssertionError(f"`batch_shape` should not contain negative values. Got {(batch_shape)}.")
```

File: kornia/augmentation/random_generator/random_generator.py [hunk 20]
```diff
@@ -883,7 +1987,7 @@ def random_sharpness_generator(
 ) -> Dict[str, torch.Tensor]:
     r"""Generate random sharpness parameters for a batch of images.
 
-    Args:
+    Args:0
```
Comment (edgarriba) [inline kornia/augmentation/random_generator/random_generator.py:1990]:
```suggestion
    Args:
```

**Review by edgarriba — COMMENTED** [review]
Few things:
- strings we should stick to double ""
- Coverage lowers quite a bit in this PR
- try also to lower deep source errors (too many)
- would be great if the random samplers can be customized

**Comment by shijianjian** [comment]
> * would be great if the random samplers can be customized

This PR only refactored the whole code base, and allows easier customization later. As a backlog, some ideas like:
- Customizing distribution samplers.
- make geometric transformations shape-agnostic.
- Easier lambda augmentations.
- Refactor mix augmentations (removing the label) for other data types like masks, key points. Improve its loss computation.
- More 3D augmentations.

File: kornia/augmentation/random_generator/random_generator.py (hunk 18 above)
Comment (edgarriba) [reply kornia/augmentation/random_generator/random_generator.py:871]: worth to mention that in docs or add  an example (or and later PR)

**Review by edgarriba — APPROVED** [review]

**Pull request merged**